### ISSUE TITLE
fn docker/dind process management should probably have a init process/service

### ISSUE THREAD

**Issue by skinowski** [fnproject/fn#689]
Currently, dind entry works like this:

```
dockerd &
sleep 3
./fn
```

Instead of this, we need a management system that is lightweight yet that can establish these dependencies clearly and robustly. (eg. if dockerd dies also kill fn server, spawn fn server only after dockerd is started, etc.)

Also see sidecar pattern and/or other ways of restructuring this. Perhaps these should be two containers where fn server runs as a sidecar container. Or adopt an init process to manage it, eg. see: http://without-systemd.org/wiki/index.php/Init

During testing, ran into a case of crashed dockerd/containerd and fn server continued running unable to service requests.

**Comment by rdallman** [comment]
there is some thinking that went into this:

in the case that `fn` dies, the container will die. in the case that docker dies, `fn` has a chance to finish any outstanding tasks. `supervisord` and friends (init) were not great solutions to this (restarting `fn` or `dockerd`) as in most cases in prior experiences, when docker dies, there is no restarting docker to bring it back. when `fn` dies, as far as state tracking and stuff kind of all bets are off, there are some cases like panic where a quick restart almost makes sense but since we lost our pipes to the user containers it doesn't really do us a whole lot of good to restart in container vs the container restarting.  the best solution to date for this was when docker dies, stop running tasks on that server, let it finish all tasks it was running, and take it out of commission. in a previous life we started containers every task and that would simply fail (in a manner safe to retry elsewhere), with hot functions we no longer have this duty and perhaps we need a higher level process running around making sure docker is not dead and shutting down the incoming task stream when it is -- the experience here is that this is also finicky (checking docker ping) but if we relax it enough we should be able to at least have an improvement over never checking. anyway, point being, this has been easier to do operationally in the past.

**Comment by skinowski** [comment]
Low priority, but another related issue is if docker dies during shutdown, fn server keeps retrying a non-existing docker socket and does not shut down.

**Comment by skinowski** [comment]
Consistently getting this:

FN started via:

`
docker run --rm --privileged -it -e NO_PROXY -e HTTP_PROXY -e FN_LOG_LEVEL=debug -e FN_ZIPKIN_URL= -e FN_DB_URL=sqlite3:///app/data/fn.db -v /home/tceylan/workdir/go/src/github.com/fnproject/fn/data:/app/data -p 8080:8080 fnproject/fnserver
`

CTRL-C results in dockerd dead, but fn running (further CTRL-C are ignored):

`
^CINFO[0290] Halting...                                   
INFO[2018-01-18T02:55:21.354945110Z] Processing signal 'interrupt'                
INFO[0290] server stopped                               
INFO[0290] hot function terminated                       app=fn-http-func cpus=200m error="<nil>" format=http id=01C43K26N847WPT00000000000 idle_timeout=20 image="tceylan/fn-http-func:0.0.14" memory=128 route=/fn-http-func
WARN[0290] docker temporary error, retrying              error="Delete http://unix.sock/containers/01C43K26N847WPT00000000000?force=1&v=1: dial unix /var/run/docker.sock: connect: no such file or directory"
WARN[0290] docker temporary error, retrying              error="Delete http://unix.sock/containers/01C43K26N847WPT00000000000?force=1&v=1: dial unix /var/run/docker.sock: connect: no such file or directory"
WARN[0290] docker temporary error, retrying              error="Delete http://unix.sock/containers/01C43K26N847WPT00000000000?force=1&v=1: dial unix /var/run/docker.sock: connect: no such file or directory"
WARN[0290] docker temporary error, retrying              error="Delete http://unix.sock/containers/01C43K26N847WPT00000000000?force=1&v=1: dial unix /var/run/docker.sock: connect: no such file or directory"
`

where

`
root     12600 12585  2 18:50 ?        00:00:16 ./fnserver
root     12645 12600  2 18:50 ?        00:00:16 [dockerd] <defunct>
`

Looks like fn server (now the init process) should honor init responsibility and reap/wait for children as well.

**Comment by rdallman** [comment]
is the SIGINT being forward to dockerd? maybe we just need to prevent that? (not sure if solely configuration issue, i.e. using docker cli in terminal) otherwise ~looks like safe shutdown is behaving as intended it's just weird that docker dies [in this configuration].

**Comment by skinowski** [comment]
I think in this case, since docker is run (as in Makefile) with interactive and with tty (-ti) so CTRL-C SIGINT delivered to both docker and fn server. We could prevent SIGINT to docker in the entry shell script.

**Comment by skinowski** [comment]
Related #771

**Comment by skinowski** [comment]
Is it time to close this? A few PRs went in, which remedies most of these issues:

1) we've just added dockerd monitoring/pid to prometheus/opencensus. #837 
2) track/consume SIGCHLD #771 
3) prevent passing SIGINT to dockerd #771 

(Also related CLI related dind by-pass: https://github.com/fnproject/cli/pull/147)

**Comment by rdallman** [comment]
+1
